cpp/dolfinx/fem/interpolate.h

```diff
@@ -1133,56 +1055,94 @@ create_nonmatching_meshes_interpolation_data(
-  std::vector<std::int32_t> cells(num_cells, 0);
-  std::iota(cells.begin(), cells.end(), 0);
-  return create_nonmatching_meshes_interpolation_data(
-      mesh0.geometry(), element0, mesh1, cells, padding);
+  auto element_u = u.function_space()->element();
+  assert(element_u);
+  const std::size_t value_size = u.function_space()->value_size();
+
+  auto& dest_ranks = interpolation_data.src_owner;
+  auto& src_ranks = interpolation_data.dest_owners;
+  auto& recv_points = interpolation_data.dest_points;
+  auto& evaluation_cells = interpolation_data.dest_cells;
+
+  // Evaluate the interpolating function where possible
+  std::vector<T> send_values(recv_points.size() / 3 * value_size);
+  v.eval(recv_points, {recv_points.size() / 3, (std::size_t)3},
+         evaluation_cells, send_values, {recv_points.size() / 3, value_size});
+
+  using dextents2 = MDSPAN_IMPL_STANDARD_NAMESPACE::dextents<std::size_t, 2>;
+
+  // Send values back to owning process
+  std::vector<T> values_b(dest_ranks.size() * value_size);
+  MDSPAN_IMPL_STANDARD_NAMESPACE::mdspan<const T, dextents2> _send_values(
+      send_values.data(), src_ranks.size(), value_size);
+  impl::scatter_values(comm, src_ranks, dest_ranks, _send_values,
+                       std::span(values_b));
+
+  // Transpose received data
+  MDSPAN_IMPL_STANDARD_NAMESPACE::mdspan<const T, dextents2> values(
+      values_b.data(), dest_ranks.size(), value_size);
+  std::vector<T> valuesT_b(value_size * dest_ranks.size());
+  MDSPAN_IMPL_STANDARD_NAMESPACE::mdspan<T, dextents2> valuesT(
+      valuesT_b.data(), value_size, dest_ranks.size());
+  for (std::size_t i = 0; i < values.extent(0); ++i)
+    for (std::size_t j = 0; j < values.extent(1); ++j)
+      valuesT(j, i) = values(i, j);
+
+  // Call local interpolation operator
+  fem::interpolate<T>(u, valuesT_b, {valuesT.extent(0), valuesT.extent(1)},
+                      cells);
 }
 
-/// @brief Interpolate from one finite element Function to another one.
-/// @param[out] u1 The function to interpolate into
-/// @param[in] u0 The function to be interpolated
-/// @param[in] cells1 List of cell indices associated with the mesh of `u1` that
-/// will be interpolated onto
-/// @param[in] cell_map Mapping of cells in mesh associated with `u1` to cells
-/// associated with `u0`
-/// @param[in] nmm_interpolation_data Auxiliary data to interpolate on
-/// nonmatching meshes. This data can be generated with
-/// create_nonmatching_meshes_interpolation_data (optional).
+/// @brief Interpolate from one finite element fem::Function to another
+/// on the same mesh.
+///
+/// Functions may be defined on 'sub-meshes'.
+///
+/// @param[out] u1 Function to interpolate into.
+/// @param[in] u0 Function to b interpolated from.
+/// @param[in] cells1 Cell indices associated with the mesh of `u1` that
+/// will be interpolated onto.
+/// @param[in] cell_map For cell `i` in the mesh associated with `u1`,
```

Comment:
@jorgensd I think this would be a lot simpler if `cell_map` was `cells0`. For each cell in `cells1` (which are in `mesh1`, `cells0` is the corresponding cell in `mesh0`.

Reply:
It would also mean that the handling of "no" cell-map (functions live on the same mesh) can be confused with having no cells on the current process that you want to interpolate into. It would mean that for every interpolation, a user would have to specify what cells they are interpolating from (major API breakage).

Reply:
> I do not agree, as we use `entity_maps` in the form creation for mesh view assembly. Having multiple concepts floating around in the different parts of the SubMesh/Meshview code is confusing to the user.

And I don't agree. Functions should take the minimum reasonable data and not more. `cell_map` as it stands could be a large data structure that the user may wish to discard. Internally the function computes what I have suggested what should be passed in. The proposal therefore can save memory and increases re-use.

Reply:
The abstraction is one level further out (in `Function.h`). I've moved the cell_map interface there.